### PR TITLE
gh-102130: Support tab completion in cmd for Libedit.

### PR DESCRIPTION
**Pull request by Constantin1489**
After I dug the rabbit hole, I wanted to share something and commit it to CPython.

My commit allows tab completion in cmd for Libedit. (**EDIT AFTER MERGE: this allows tab completion for pdb with libedit**)

`tab` of `readline.parse_and_bind("tab: complete")` is a symbolic character name. Libedit doesn't support a number of symbolic character names(like TAB, ESC, DEL). 

FYI, for better compatibility, Python Documents should suggest using backslashed escape sequences(e.g. `\t`, `\a`) or the ASCII character corresponding to the octal number(or octal excape sequences).(e.g. `\011`),  It's because the only thing GNU readline(`man bash` and `/Readline Key Bindings` to search the part.) and Libedit(`man editrc`) have in common. 

Also, unlike GNU emacs software,  GNU readline doesn't support control characters of the form '^character' (e.g. '^I' it's a tab). 
Both GNU readline(`man bash` and `/Readline Key Bindings` to search the part.) and Libedit(`man editrc`) support the following backslashed escape sequences and octal escape sequences.

```
\a     bell
\b     backspace
\f     form feed
\n     newline
\r     carriage return
\t     horizontal tab
\v     vertical tab
\nnn   octal escape sequences
```
GNU readline's predefined commands are in the '/Readline Command Names' of `man bash`.
Libedit's predefined commands are in the section "EDITOR COMMANDS" of `man editrc`. But there are other undocumented commands in its source code.
Or you can get a list by `import readline; readline.parse_and_bind('bind -l')` (Libedit only. This shows undocumented commands too).

* Issue: gh-102130

----
:books: Documentation preview :books:: https://cpython-previews--107748.org.readthedocs.build/

### CODE REVIEW

**Comment by Constantin1489** [comment]
**There is something that Python core developers decide.**

1. Supporting "TAB" symbolic character name for libedit in "cmd.py".

If this is supported, then other third-party libraries can support Libedit tab completion.
I don't want this support because it will give a false impression of common key sequences and the requirement to support other symbolic character names.

```
readline_doc = getattr(readline, '__doc__', '')
if readline_doc is not None and 'libedit' in readline_doc: 
    if self.completekey is 'tab':
        self.completekey = r'"\t"'
```
OR just

```
if self.completekey is 'tab':
    self.completekey = r'"\t"'
```

2. Removing editrc examples in readline.rst.

Because of the absence of the GNU inputrc example in `readline.rst`, the editrc example may be just for hotfix to do a tab complete in cmd or pdb.

For Vim keybinding in the editrc example(`python:bind -v`), I think it's valid. For this reason, Vim keybinding for GNU inputrc example is needed.

3. Changing keybindings in related test files.
There are `^I` and `tab` in related test files.  If you need me to change them, I will do it.

** Just so you know, if you want to debug this problem, please check what I found below.**

Use one of these lines to enable tab completion in Python using `~/.editrc` for libedit. 
```
python:bind "\t" rl_complete
python:bind "\011" rl_complete
python:bind \\t rl_complete
python:bind "	" rl_complete
# "	" is a tab.
```

To use them with`readline.parse_and_bind()` in Python, remove `'python:'` prefix. To make clear, If you did `import readline; readline.parse_and_bind('bind "\t" rl_complete')` in pdb interactively (e.g., `$python -m pdb somefile.py`), it will autocomplete file names of the current directory. It's normal because the binding statement wasn't called in the pdb library.

The key sequences(`"\t"`, `"\011"`) of the first two lines in the code block above are documented in `man editrc`. The first one is called a backslashed escape sequence. The second one is called the ASCII character corresponding to the octal number(or octal escape character).

But the last two lines are not. I stumble across them.

To do another complete command in Libedit, this is possible.
``import readline; readline.parse_and_bind('bind -s "\011" ee')``
In this case, If you press a tab, 'ee' will be typed.
Other possible options are appeared in `man editrc`.

I found several ways to debug it. (this is an example.)

1. In Python interpreter, `import readline; readline.parse_and_bind('bind -s "\011" ee')`
2. Replace `'tab'` with `''` in the pdb.py and cmd.py of your python directory in `$PATH`. e.g. `/usr/local/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/pdb.py`. 

**How to check the keybinding setting applied with Libedit.**

```
import readline
readline.parse_and_bind('bind')
```
This shows the user's current applied libedit readline setting.

File: Doc/library/readline.rst
Comment (Constantin1489) [inline Doc/library/readline.rst]: I don't have a clue how to turn off a syntax highlight.

**Comment by gaogaotiantian** [comment]
I don't think we should change the default and hope the completekey passed in is correct. It could break users code, for example, if they compare `self.completekey == tab`. Also, if the user subclassed the `cmd.Cmd` class and did something `pdb` does like

```python
class MyCmd(cmd.Cmd):
    def __init__(self, completekey='tab'):
        cmd.Cmd.__init__(self, completekey)
```

They will still have the issue.

We should convert the completekey internally and just do a different binding command with `libedit`. Minimal code changes and fix things for good.

Also, why do you want to change the `site.py` file? It works for now I believe. Python REPL has the complete feature with `libedit`.

I believe the right way to do this is just add an extra check in `cmd.Cmd` for `libedit`, and do the correct thing (convert `tab` to `"\t"` or the equivalent in `libedit`, then use a different binding command).

Also, if you want to finish this, you'd need a regression test on at least `cmd.Cmd`, it would be easier after #110945 is merged so you'll have a helper function for pty.

I can help you with the review and find the core-dev for approval if you want to work on this.

**Comment by Constantin1489** [comment]
Yes, what you pointed it is valid. 
So in this case, you want Python should change the tab internally for portability. 

>  Minimal code changes and fix things for good.

Because I was unsure what the decision I made. This is what I asked in number 1 https://github.com/python/cpython/pull/107748#issuecomment-1668593509.
In that case, a user who uses libedit with your custom object will experience when press 't' will be ignored sometimes. 

At the time I wrote this PR, I thought, that if a user did something wrong, the wrong result must be thrown. And the docs must give a  pledge. (You know when learning the linux thing there must be some reason I don't know and some result is unintuitive or counterintuitive for newcomer.). 

Thank you for giving me direction for what I consider. I will do that within a week.

The second thing you mention(site.py) is just for a tone and manner thing(second and third thing that I asked in https://github.com/python/cpython/pull/107748#issuecomment-1668593509). I will revert it.

Thank you for helping me.

**Comment by gaogaotiantian** [comment]
I believe you only need to change `cmd.py`, which would make this PR compact and easy to review. I'd recommend you to wait for #110945 so you'll have a test framework for completion. Actually enabling the test in that PR on libedit would just do.

BTW, you should not use `is` to compare strings. In this very specific case it actually works because of the python internal cache for intern strings, but that's definitely not what you should rely on. Always use `==` to check the string equality.

File: Lib/cmd.py
Comment (gaogaotiantian) [inline Lib/cmd.py]:
```suggestion
                if 'libedit' in getattr(readline, '__doc__', ''):
```

**Comment by Constantin1489** [comment]
@gaogaotiantian I'm afraid to say this but, I did a test with **python3.11.6**, but libedit expansion works(which means expands file lists in PWD like `ls` command), but it doesn't expand the pdb command. 

When I tested this PR with python3.11.4, previous PR and changed PR worked and expanded pdb commands correctly.

`git log v3.11.4..v3.11.6 Lib/pdb.py` shows 43ff8fcce21377cdce3a594ea10e1a045bb70752 commit. And I can confirm that reverting this recovers pdb expansion.

**Comment by gaogaotiantian** [comment]
I have the hardest time believing that commit can cause the completion issue. The code should not be executed at all for completion. Are you sure you did the test correctly?

**Comment by Constantin1489** [comment]
I'm sorry. You are right. Maybe my environment goes something wrong.

**Comment by gaogaotiantian** [comment]
Hi @Constantin1489 , now that https://github.com/python/cpython/pull/111826 is merged, you have the utility to write tests for the feature. Could you:

* Add a test in `test_cmd.py` to do some basic testing for command completion? It does not need to be super fancy, refer to https://github.com/python/cpython/pull/111826 for the usage of `run_pty()`. 
* Remove the check for `libedit` in https://github.com/python/cpython/pull/111826 so it would be another confirmation that the feature works with `libedit`? Do not remove the `import_module` piece because that's the magic to skip the test if `readline` is not available.

**Comment by Constantin1489** [comment]
Thank you, @gaogaotiantian !

**Comment by gaogaotiantian** [comment]
Ah, that actually reveals a deeper issue for libedit compatibility. It's out of the scope of this PR, so could you revert the changes to `test_pdb`? I'll deal with it later.

Also, for the cmd test, could you change the `do_` function to something other than `do_EOF`? `do_EOF` has a very special meaning and we should not test with it if it's not required - that could confuse people. Just make up a command, print something, do the completion and checks for the output. I don't think you need the `hello!` part.

If you do want to test the help part, you need to explain it more. It's not obvious to most people that `\t\t` brings up the help page - listing "everything" is not directly linked to "help" being displayed.

**Comment by Constantin1489** [comment]
1. Okay!
2. Thank you! I refined the test and made it intuitive.
3. I reconsidered it and noticed that it requires a prior knowledge which is not that needed. So I decided to check a tab completion for an exact one command.

**Comment by gaogaotiantian** [comment]
Yeah the test is simple and straightforward now!

File: Misc/NEWS.d/next/Library/2023-08-07-21-11-24.gh-issue-102130._UyI5i.rst
Comment (gaogaotiantian) [inline Misc/NEWS.d/next/Library/2023-08-07-21-11-24.gh-issue-102130._UyI5i.rst]:
```suggestion
Support tab completion in :mod:`cmd` for Libedit.
```

**Review by gaogaotiantian — APPROVED** [review]

**Comment by gaogaotiantian** [comment]
Hi @encukou , could you take a look at this as you just reviewed my related code? This is to enable the automatic conversion in `cmd.Cmd` for libedit.

File: Lib/cmd.py
Comment (encukou) [inline Lib/cmd.py]: I don't think we want the docstring to become API, but it's getting used in too many places now. Would either you mind exposing a dedicated attribute, like `readline.backend`?
(Preferably in a separate PR, which doesn't need to block this one.)

**Comment by encukou** [comment]
> I don't think we should change the default and hope the completekey passed in is correct. It could break users code

Have you considered changing the default `completekey` in `Cmd.__init__` to depend on the readline backend? That sounds like it would help more users. Since the current default is essentially unusable with libedit, it shouldn't break people?

> Also, if the user subclassed the cmd.Cmd class and did something pdb does like
>```
> class MyCmd(cmd.Cmd):
>     def __init__(self, completekey='tab'):
>         cmd.Cmd.__init__(self, completekey)
> ```

Yes, that's incompatible with libedit -- but IMO, having to change such code to support a new backend is OK. To make this easier, `Cmd.__init__` could take `None` and convert it to the appropriate default.

File: Lib/cmd.py
Comment (gaogaotiantian) [reply Lib/cmd.py]: I can work on that, but yes in a separate PR because we want to replace all the usages as such in the PR as well.

**Comment by gaogaotiantian** [comment]
> Have you considered changing the default `completekey` in `Cmd.__init__` to depend on the readline backend? That sounds like it would help more users. Since the current default is essentially unusable with libedit, it shouldn't break people?

We can definitely change the default `completekey in `Cmd.__init__`, but I think we should do the mapping anyway

> Yes, that's incompatible with libedit -- but IMO, having to change such code to support a new backend is OK. To make this easier, `Cmd.__init__` could take `None` and convert it to the appropriate default.

Yes it's not the end of the world if our "new supported platform" requires changes from the users. However, if we *can* make it work smoothly for them, why not? There is a non-trivial number of [code](https://github.com/search?q=completekey%3D%27tab%27+language%3APython&type=code&l=Python) that uses `completekey='tab'` and we can make all of them work magically with libedit, without them knowing about it.

There are some more reasons that I'd like to keep the mapping:
1. `'tab'` is much more easier to use and understand, than `None` representing a tab, or `r'"\t"'` (did you notice the `r` prefix and the double quote in the string?).
2. There could be other keys that have different representation in libedit and GNU readline. Yes they are relatively rare to use, but who knows? We can easily add mappings for them to support both platforms and prevent the users from dealing with them. Otherwise, they need to figure out the backend on their own, and choose the correct term.

So, overall, I think the current implementation is better than having `None` as default and asking the users to do their magic.

**Comment by encukou** [comment]
I'm not 100% sure about this, but it seems possible to configure libedit so that `parse_and_bind("bind tab rl_complete")` is valid, and `tab` doesn't mean `\t`.
If that's the case, then I think we must not second-guess the user. (And if a plausible future version of libedit could make it possible, the same applies.)

>  we can make all of them work magically with libedit, without them knowing about it

Be careful with magic. It often has unexpected side effects.
Python's stdlib is so widely used, it must be more risk-averse than most projects.

> 'tab' is much more easier to use and understand, than None representing a tab

Think of it as `None` representing *the default*, which is a well-established convention.

> There could be other keys that have different representation in libedit and GNU readline. Yes they are relatively rare to use, but who knows? We can easily add mappings for them to support both platforms and prevent the users from dealing with them. Otherwise, they need to figure out the backend on their own, and choose the correct term.

I don't agree with this approach: open-ended scope means that the software can never really be “done”. And if small improvements do end up being added regularly, there'll be uncertainty about whether a particular version of Python supports the one you want, and how to support older versions.
Incremental improvements like this are *much* better done in a library that can be updated independently of Python.

**Comment by gaogaotiantian** [comment]
> I'm not 100% sure about this, but it seems possible to configure libedit so that `parse_and_bind("bind tab rl_complete")` is valid, and `tab` doesn't mean `\t`.
> If that's the case, then I think we must not second-guess the user. (And if a plausible future version of libedit could make it possible, the same applies.)

I agree, but I don't think that's valid after quickly looking through the code (I might miss something). Also, I don't think logically it means anything. What could a user possibly want when they do `parse_and_bind("bind tab rl_complete")`, if they don't want to bind `<tab>` to complete key? We definitely should not second guess user, but we should also be reasonable.

My understanding for the standard library is - it should be as simple as possible in the common case, yet still available in uncommon cases.

The most common case would be using `tab` as complete key. If the user simply inherit `cmd.Cmd` and does not do anything with `completekey` argument, then `None` and `'tab'` both work.

However, there are existing code like I mentioned above, which already inherits `cmd.Cmd` and copied the `completekey='tab'` argument. Their code will not work on `libedit` if we use `None`. Their code will work, if we map `'tab'` to `r'"\t"'` or `'^I'`.

Moreover, there's another not so uncommon scenario (arguably the second most common case): the user does NOT want complete feature at all. For example, they require the ability for users to input `\t` in the console (which is what needed in pdb multi-line mode). Question: what do they need to do?

If I were the user without any knowledge about `readline` (because I'm just a developer using a command line framework provided by Python), the obvious move is to pass `completekey=None`. That's super intuitive.

This is the ambiguity of using `None` as default - it only works, if passing `None` to that argument does not make much sense. In this case, passing `None` to `completekey` is obviously (to me) trying to disable the complete feature.

Passing `None` is not working beautifully under the hood now (it's binding an invalid value rather than unbinding), but it works from high level. The user will get expected result. How would a user achieve such feature, without knowing the details about `readline`?

If the user does know about `readline` and the detailed differences about `libedit` and `GNU readline`, they can always do whatever they need after `cmd.Cmd.__init__`. The code is almost identical to what they need to do to pass different complete keys.

Maybe we should not map other keys because that might be a black hole, but I do think making `'tab'` work is the best solution for now. Asking users to change `completekey='tab'` to `completekey=None` to make `tab` work, seems off to me.

**Comment by encukou** [comment]
Disabling completion altogether would be a weird feature to ask for, but, OK. You convinced me that practicality should beat purity here.
I'd rather still keep `self.completekey` to the passed-in value, though, and only do the mapping for `parse_and_bind`.

Also, could you update the docs to say that `tab` is treated specially, to help avoid surprises?

File: Lib/cmd.py
Comment (gaogaotiantian) [inline Lib/cmd.py]:
```suggestion
                    if self.completekey == 'tab':
                        # libedit expects "^I" for tab completion instead of "tab"
                        readline.parse_and_bind("bind ^I rl_complete")
                    else:
                        readline.parse_and_bind("bind "+self.completekey+" rl_complete")
```

File: Doc/library/cmd.rst
Comment (Constantin1489) [inline Doc/library/cmd.rst:35]: @gaogaotiantian Thank you for giving me a chance! I added the fact. Please, feel free to edit it!

Comment (gaogaotiantian) [reply Doc/library/cmd.rst:35]: I would suggest using a ``.. note::`` section here, which is often used for extra information.

I'll not the expert of wording, but I'd say something like

```
``'tab'`` will be treated as ``'^I'`` for ``libedit``, as that's the representation for :kbd:`Tab` in ``libedit``
```

I would ask @encukou about the wording or maybe some docs experts.

Comment (encukou) [inline Doc/library/cmd.rst]: It seems to me that this would be easier to explain after #112511 is merged:
```suggestion

   The default, ``'tab'``, is treated specially, so that it refers to the
   :kbd:`Tab` key on every :data:`readline.backend`.
   Specifically, if :data:`readline.backend` is ``libedit``,
   ``Cmd`` will use ``'^I'`` instead of ``'tab'``.
   Note that other values are not treated this way, and might only work
   with a specific backend.
```

File: Lib/cmd.py
Comment (encukou) [inline Lib/cmd.py]: Please make the lines a little shorter (PEP 8 calls for 80 characters).
Maybe something like this (untested):
```suggestion
                if 'libedit' in getattr(readline, '__doc__', ''):
                    if self.completekey == 'tab':
                        # libedit uses "^I" instead of "tab"
                        command_string = "bind ^I rl_complete"
                    else:
                        command_string = f"bind {self.completekey} rl_complete"
                else:
                    command_string = f"{self.completekey}: complete"
                readline.parse_and_bind(command_string)
```

File: Doc/library/cmd.rst
Comment (Constantin1489) [reply Doc/library/cmd.rst]: @encukou Thank you for the clear explanation!

File: Lib/cmd.py
Comment (Constantin1489) [reply Lib/cmd.py:119]:
```
>>> completekey = r'"\t"'
>>> f"bind {completekey} rl_complete" == "bind "+completekey+" rl_complete"
True
```
I'm sorry for this. I never had this kind of deep-level repo commit before... I don't know if I did it correctly...

File: Doc/library/cmd.rst
Comment (Constantin1489) [inline Doc/library/cmd.rst:31]: Because of 
```
Lib/site.py:447:        if readline.backend == 'editline':
Lib/test/test_pdb.py:3296:        if readline.backend == "editline":
Lib/test/test_readline.py:149:        self.assertIn(readline.backend, ("readline", "editline"))
Lib/test/test_readline.py:177:            if readline.backend == "editline":
Lib/test/test_readline.py:204:is_editline = readline.backend == "editline"
Lib/test/test_readline.py:22:    is_editline = readline.backend == "editline"
```

File: Lib/cmd.py
Comment (Constantin1489) [inline Lib/cmd.py:111]: Because of 
```
Lib/site.py:447:        if readline.backend == 'editline':
Lib/test/test_pdb.py:3296:        if readline.backend == "editline":
Lib/test/test_readline.py:149:        self.assertIn(readline.backend, ("readline", "editline"))
Lib/test/test_readline.py:177:            if readline.backend == "editline":
Lib/test/test_readline.py:204:is_editline = readline.backend == "editline"
Lib/test/test_readline.py:22:    is_editline = readline.backend == "editline"
```

**Comment by encukou** [comment]
And one more thing I always forget to point out (sorry for that)!

At the end of the `Cmd` class documentation (before `.. _cmd-objects:`), add a versionchanged note like
```
   .. versionchanged:: 3.11
   ``completekey='tab'`` is replaced by ``'^I'`` for ``editline``.
```

Also, please ignore the free-threaded CI failures; they're not your fault and they're allowed to fail for now.

File: Doc/library/cmd.rst
Comment (Constantin1489) [inline Doc/library/cmd.rst:46]: Thank you!! 
Because other versionchanged notes have three blank spaces in front of the content, I appended them!
```
$grep -rin versionchanged -C2
./library/exceptions.rst-756-   :py:const:`~errno.EPERM`, and :py:const:`~errno.ENOTCAPABLE`.
./library/exceptions.rst-757-
./library/exceptions.rst:758:   .. versionchanged:: 3.11.1
./library/exceptions.rst-759-      WASI's :py:const:`~errno.ENOTCAPABLE` is now mapped to
./library/exceptions.rst-760-      :exc:`PermissionError`.
```

File: Misc/NEWS.d/next/Library/2023-08-07-21-11-24.gh-issue-102130._UyI5i.rst
Comment (Constantin1489) [inline Misc/NEWS.d/next/Library/2023-08-07-21-11-24.gh-issue-102130._UyI5i.rst]: Should I change Libedit to editline of this and this commit post's name?
```
gh-102130: Support tab completion in cmd for Libedit.
->
gh-102130: Support tab completion in cmd for editline.
```

**Comment by encukou** [comment]
change the NEWS text, the commit names don't matter

On December 1, 2023 5:02:04 PM GMT+01:00, Constantin Hong ***@***.***> wrote:
>@Constantin1489 commented on this pull request.
>
>
>
>> @@ -0,0 +1 @@
>+Support tab completion in :mod:`cmd` for Libedit.
>
>Should I change Libedit to editline of this and this commit post's name?
>```
>gh-102130: Support tab completion in cmd for Libedit.
>->
>gh-102130: Support tab completion in cmd for editline.
>```
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/python/cpython/pull/107748#pullrequestreview-1760125757
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: Lib/test/test_cmd.py
Comment (encukou) [inline Lib/test/test_cmd.py]: OK! The PR now looks good to me.
However, to be extra sure, I installed editline and ran the test, and got:

```
AssertionError: b'tab_completion_test' not found in bytearray(b't\t\r\n(Cmd) t\x07\r\x1b[8Gab_completion_test\r\ntab completion success\r\n'
```

Does that not happen on your machine?
It would be nice to fix this properly -- perhaps configuring the library to not move the cursor this way -- but it's OK to leave that to the future, when more tests are added in this area.
For now, this should be enough:

```suggestion
        self.assertIn(b'ab_completion_test', output)
```

Comment (encukou) [reply Lib/test/test_cmd.py]: Are you OK with that change? If so, let's merge this!
(I'm asking because you might know a better fix :)

Comment (Constantin1489) [reply Lib/test/test_cmd.py]: In my mac, `print(output)` result is
```
./python.exe Lib/test/test_cmd.py
bytearray(b't\t\r\n(Cmd) tab_completion_test\r\ntab completion success\r\n')
.....
----------------------------------------------------------------------
Ran 5 tests in 0.068s

OK
```

Comment (Constantin1489) [reply Lib/test/test_cmd.py]: FYI, when I intentionally failed the test,
```
FAIL: test_basic_completion (__main__.CmdTestReadline.test_basic_completion)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/constantinhong/TODO/git-log-msg/cpython_libedit/cpython/Lib/test/test_cmd.py", line 290, in test_basic_completion
    self.assertIn(b'tsab_completion_test', output)
    ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: b'tsab_completion_test' not found in bytearray(b't\t\r\n(Cmd) tab_completion_test\r\ntab completion success\r\n')

----------------------------------------------------------------------
Ran 5 tests in 0.073s

FAILED (failures=1)
```

File: Doc/library/cmd.rst
Comment (Constantin1489) [inline Doc/library/cmd.rst]: I want to make sure of this... 
the `readline.backend` (https://github.com/python/cpython/pull/112511) will be added `.. versionadded:: 3.13` but this PR will support tab completion since 3.11.
So in this case, 3.11-3.12 backport will have a different code.. But, this PR's code uses the new backend API.

Comment (encukou) [reply Doc/library/cmd.rst]: Oh, sorry! I missed the number.
It definitely won't be changed in `3.11.0` or `3.12.0`, which were already released, so this does need to be `3.13`.
I'm not convinced that this should be backported to earlier versions (3.11.7 and 3.12.2), but, that can be discussed on the issue.
The backports would need to adapt to the missing API.
```suggestion
   .. versionchanged:: 3.13
```

**Comment by encukou** [comment]
Thank you for your contribution!

**Comment by tanloong** [comment]
Hi @gaogaotiantian, as libedit is now supported, is this part in `test_pdb.py` not needed anymore?  https://github.com/python/cpython/blob/main/Lib/test/test_pdb.py#L4780-L4781

Can I make a PR to remove it?

**Comment by gaogaotiantian** [comment]
@tanloong, I don't believe you can. If you read the full discussion (to be more specific - https://github.com/python/cpython/pull/107748#issuecomment-1811780154), you'll find that we tried to do that but failed. The problem is `libedit` has some issues for inputing spaces so all the multiline tests would fail. Feel free to give it a try (you'll need libedit so make sure `readline.backend` returns `editline` before you start working on it). If you can solve the problem, I'd be happy to review the PR.

**Comment by tanloong** [comment]
@gaogaotiantian, thanks so much for clarifying, I totally missed it from not reading carefully.
 
I want to be honest I'm not sure if I can fix the libedit space issue, but I'll still give it a try. Thanks again!

**Pull request merged**
